### ISSUE TITLE
GrainInterfaceMap sent to the client should be streamlined

### ISSUE THREAD

**Issue by jdom** [dotnet/orleans#3138]
Currently when the client connects to a silo, it receives a `GrainInterfaceMap` with a lot of information about the grain implementations in the silo. This is the same information that silos pass to each other, and contains stuff like the placement strategies and so on.
The client does not care much about that information, other than some basic stuff to know the type code that it should send its messages too (from the interface type, plus the grain class prefix it needs to determine a type code).
Well, the point is that because it's receiving all of this extra information that it doesn't really care about, in some circumstances the client will fail to deserialize this map, and fail to connect entirely, unless you start hacking around with assembly loading.
This is a bigger issue since we opened up extensibility to provide your own custom placement strategies that might be in DLLs that the client either doesn't have access to or even if it does, it doesn't eagerly load to make sure the serializer is registered before it gets an instance of that type.
This can only get worse as we open extensibility more and more (such as with versioning, activation strategies, heterogeneous silos, multi clusters, etc).

I suggest that for sending information from silo to clients, we use a separate streamlined structure with just the information that clients need.

For example, see an exception that we had to work around in our tests:

```text
MESSAGE:
System.TypeAccessException : Named type "TestVersionGrains.VersionAwarePlacementStrategy" is invalid: Type string "TestVersionGrains.VersionAwarePlacementStrategy" cannot be resolved.
+++++++++++++++++++
STACK TRACE:
at Orleans.Serialization.BinaryTokenStreamReader.ReadSpecifiedTypeHeader(SerializationManager serializationManager) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Serialization\BinaryTokenStreamReader.cs:line 940
at Orleans.Serialization.SerializationManager.DeserializeInner(Type expected, IDeserializationContext context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Serialization\SerializationManager.cs:line 1535
at Orleans.OrleansCodeGenOrleans_Runtime_GrainClassDataSerializer.Deserializer(Type expected, IDeserializationContext context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\obj\Release\Orleans.orleans.g.cs:line 2282
at Orleans.Serialization.SerializationManager.DeserializeInner(Type expected, IDeserializationContext context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Serialization\SerializationManager.cs:line 1574
at Orleans.Serialization.BuiltInTypes.DeserializeDictionary[K,V](Type expected, IDeserializationContext context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Serialization\BuiltInTypes.cs:line 575
at Orleans.Serialization.SerializationManager.DeserializeInner(Type expected, IDeserializationContext context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Serialization\SerializationManager.cs:line 1574
at Orleans.OrleansCodeGenOrleans_Runtime_GrainInterfaceMapSerializer.Deserializer(Type expected, IDeserializationContext context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\obj\Release\Orleans.orleans.g.cs:line 2215
at Orleans.Serialization.SerializationManager.DeserializeInner(Type expected, IDeserializationContext context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Serialization\SerializationManager.cs:line 1574
at Orleans.Serialization.BuiltInTypes.DeserializeOrleansResponse(Type expected, IDeserializationContext context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Serialization\BuiltInTypes.cs:line 1927
at Orleans.Serialization.SerializationManager.DeserializeInner(Type expected, IDeserializationContext context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Serialization\SerializationManager.cs:line 1574
at Orleans.Serialization.SerializationManager.Deserialize(Type t, BinaryTokenStreamReader stream) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Serialization\SerializationManager.cs:line 1463
at Orleans.Runtime.Message.DeserializeBody(SerializationManager serializationManager, List`1 bytes) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Messaging\Message.cs:line 429
at Orleans.Runtime.Message.GetDeserializedBody(SerializationManager serializationManager) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Messaging\Message.cs:line 389
at Orleans.Runtime.GrainReference.ResponseCallback(Message message, TaskCompletionSource`1 context) in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Runtime\GrainReference.cs:line 425 
--- End of stack trace from previous location where exception was thrown --- 
at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at Orleans.OutsideRuntimeClient.<StartInternal>d__64.MoveNext() in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Runtime\OutsideRuntimeClient.cs:line 277 
--- End of stack trace from previous location where exception was thrown --- 
at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at Orleans.OutsideRuntimeClient.<Start>d__63.MoveNext() in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Runtime\OutsideRuntimeClient.cs:line 242 
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at Orleans.ClusterClient.<Connect>d__22.MoveNext() in D:\j\workspace\netfx-bvt_prtest437995ca\src\Orleans\Core\ClusterClient.cs:line 105
```

**Comment by yevhen** [comment]
+1 I've ran into the issue described above while deploying code using custom placement strategy

**Comment by benjaminpetit** [comment]
Fixed with #3886
